src/algebra/gcd_monoid.lean

```diff
@@ -300,6 +300,44 @@ dvd_antisymm_of_normalize_eq (normalize_gcd _ _) (normalize_gcd _ _)
   (gcd_dvd_gcd (dvd_refl _) (dvd_of_associated h))
   (gcd_dvd_gcd (dvd_refl _) (dvd_of_associated h.symm))
 
+/-- Represent a divisor of `m * n` as a product of a divisor of `m` and a divisor of `n`. -/
+noncomputable def prod_dvd_and_dvd_of_dvd_prod {m n k : α} (H : k ∣ m * n) :
+  { d : {m' // m' ∣ m} × {n' // n' ∣ n} // k = d.1 * d.2 } :=
+begin
+  by_cases h0 : gcd k m = 0,
+  { rw gcd_eq_zero_iff at h0,
+    rcases h0 with ⟨rfl, rfl⟩,
+    refine ⟨⟨⟨0, dvd_refl 0⟩, ⟨n, dvd_refl n⟩⟩, _⟩,
+    simp },
+  { let a := classical.some (gcd_dvd_left k m),
+    refine ⟨⟨⟨gcd k m, gcd_dvd_right _ _⟩, ⟨a, _⟩⟩, classical.some_spec (gcd_dvd_left k m)⟩,
+    suffices h : gcd k m * a ∣ gcd k m * n,
+    { cases h with b hb,
+      use b,
+      rw mul_assoc at hb,
+      apply mul_left_cancel' h0 hb },
+    rw ← classical.some_spec (gcd_dvd_left k m),
+    transitivity gcd k m * normalize n,
+    { rw ← gcd_mul_right,
+      exact dvd_gcd (dvd_mul_right _ _) H },
+    { apply dvd.intro ↑(norm_unit n)⁻¹,
+      rw [normalize_apply, mul_assoc, mul_assoc, ← units.coe_mul],
+      simp } }
+end
```

Comment:
Is there a reason you used a subtype rather than exists? It seems like the following requires a few less brackets:

```suggestion
lemma exists_dvd_and_dvd_of_dvd_mul {m n k : α} (H : k ∣ m * n) :
  ∃ d₁ (hd₁ : d₁ ∣ m) d₂ (hd₂ : d₂ ∣ n), k = d₁ * d₂ :=
begin
  by_cases h0 : gcd k m = 0,
  { rw gcd_eq_zero_iff at h0,
    rcases h0 with ⟨rfl, rfl⟩,
    refine ⟨0, dvd_refl 0, n, dvd_refl n, _⟩,
    simp },
  { obtain ⟨a, ha⟩ := gcd_dvd_left k m,
    refine ⟨gcd k m, gcd_dvd_right _ _, a, _, ha⟩,
    suffices h : gcd k m * a ∣ gcd k m * n,
    { cases h with b hb,
      use b,
      rw mul_assoc at hb,
      apply mul_left_cancel' h0 hb },
    rw ← ha,
    transitivity gcd k m * normalize n,
    { rw ← gcd_mul_right,
      exact dvd_gcd (dvd_mul_right _ _) H },
    { apply dvd.intro ↑(norm_unit n)⁻¹,
      rw [normalize_apply, mul_assoc, mul_assoc, ← units.coe_mul],
      simp } }
end
```

Reply:
All of this started with copy-pasting `nat.prod_dvd_and_dvd_of_dvd_prod`, so I just copied that name and format.

Reply:
I prefer exists, though.